File: ts-ast/operations/op_unary.ts
import { CompositeGeneratorNode as CGN } from 'langium/generate';

import { jsonObject, jsonString } from '../utils/json.js';
import type { BaseValue } from './base.js';
import { OpNode } from './codegen.js';
import type { UnaryOperator } from './op_operator.js';

export class UnaryOp extends OpNode {
    readonly op = 'unary' as const;

    constructor(
        readonly ope: UnaryOperator,
        readonly value: BaseValue,
    ) {
        super();
    }

    override codegen(): CGN {
        return jsonObject([
            ['op', jsonString(this.op)],
            ['ope', jsonString(this.ope)],
            ['value', this.value.codegen()],
        ]);
    }
}

File: src/jsonmlir/operations/op_unary.py
from __future__ import annotations

from collections.abc import Sequence
from enum import Enum
from typing import TYPE_CHECKING, Any, Literal

from jsonmlir.operations.codegen import OpNode
from jsonmlir.operations.op_binary import BinaryOp
from jsonmlir.operations.op_constant import ConstOp
from jsonmlir.operations.op_operator import OperatorOp
from jsonmlir.utils.enum_scalars import Scalar
from jsonmlir.utils.trace import trace_step
from jsonmlir.variables.val.val import ValNode

if TYPE_CHECKING:
    from jsonmlir.operations.base import BaseValue


class UnaryOperator(Enum):
    """Unary operators supported by :class:`UnaryOp`."""

    negOp = "-"
    negFOp = "-f"
    notOp = "!"


class UnaryOp(OpNode):
    """Apply a unary operator, lowered to a binary operation with a constant."""

    op: Literal["unary"] = "unary"
    ope: UnaryOperator
    value: BaseValue

    @trace_step("UnaryOp: {self.ope.value}")
    def codegen(self) -> Sequence[ValNode[Any]]:
        match self.ope:
            # -f x  ->  x *f -1.0
            case UnaryOperator.negFOp:
                return BinaryOp(
                    lhs = self.value,
                    rhs = ConstOp(val=-1.0, type=Scalar.f64),
                    ope = OperatorOp.timesFOp
                ).codegen()
            # -  x  ->  x  * -1
            case UnaryOperator.negOp:
                return BinaryOp(
                    lhs = self.value,
                    rhs = ConstOp(val=-1, type=Scalar.i64),
                    ope = OperatorOp.timesOp
                ).codegen()
            # !x    ->  x xor true
            case UnaryOperator.notOp:
                return BinaryOp(
                    lhs = self.value,
                    rhs = ConstOp(val=1, type=Scalar.i1),
                    ope = OperatorOp.xorOp
                ).codegen()
